From 8aa7de6284b96f29911baa3afb438826ba44001b Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Fri, 14 Sep 2018 01:03:27 +0200
Subject: [PATCH 11/11] Check for envvar to actually enable PBA instead of the
 opposite

Signed-off-by: Rob Walker <bob.mt.wya@gmail.com>
---
 dlls/wined3d/device.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 782628b9d8..1aea4ecd79 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -824,17 +824,13 @@ static void create_buffer_heap(struct wined3d_device *device, struct wined3d_con
 {
     const struct wined3d_gl_info *gl_info = &device->adapter->gl_info;
     BOOL use_pba = FALSE;
-    char *env_pba_disable;
+    char *env_pba_enable;
 
     if (!gl_info->supported[ARB_BUFFER_STORAGE])
     {
         FIXME_(d3d_perf)("Not using PBA, ARB_buffer_storage unsupported.\n");
     }
-    else if ((env_pba_disable = getenv("PBA_DISABLE")) && *env_pba_disable != '0')
-    {
-        FIXME("Not using PBA, envvar 'PBA_DISABLE' set.\n");
-    }
-    else
+    else if ((env_pba_enable = getenv("PBA_ENABLE")) && *env_pba_enable != '0')
     {
         //(Firerat) is it worth initialising an int for vram?
         unsigned int vram_mb = device->adapter->vram_bytes / 1048576;
@@ -908,6 +904,10 @@ static void create_buffer_heap(struct wined3d_device *device, struct wined3d_con
 
         use_pba = TRUE;
     }
+    else
+    {
+        FIXME("Not using PBA, envvar 'PBA_ENABLE' not set.\n");
+    }
 
 fail:
     device->use_pba = use_pba;
-- 
2.19.1

