diff --git a/dlls/wined3d/context.c b/dlls/wined3d/context.c
index b513705874d..7e4f52d2d13 100644
--- a/dlls/wined3d/context.c
+++ b/dlls/wined3d/context.c
@@ -4991,7 +4991,11 @@ void draw_primitive(struct wined3d_device *device, const struct wined3d_state *s
     if (parameters->indexed)
     {
         struct wined3d_buffer *index_buffer = state->index_buffer;
-        if (!index_buffer->buffer_object || !stream_info->all_vbo)
+        if (index_buffer->cs_persistent_map)
+        {
+            idx_data = index_buffer->cs_persistent_map->range.offset;
+        }
+        else if (!index_buffer->buffer_object || !stream_info->all_vbo)
         {
             idx_data = index_buffer->resource.heap_memory;
         }
